hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/ZookeeperBasedLockProvider.java

```diff
@@ -0,0 +1,161 @@
+    checkRequiredProps(lockConfiguration);
+    this.lockConfiguration = lockConfiguration;
+    this.curatorFrameworkClient = CuratorFrameworkFactory.builder()
+        .connectString(lockConfiguration.getConfig().getString(ZK_CONNECT_URL_PROP))
+        .retryPolicy(new BoundedExponentialBackoffRetry(lockConfiguration.getConfig().getInteger(LOCK_ACQUIRE_RETRY_WAIT_TIME_IN_MILLIS_PROP),
+            5000, lockConfiguration.getConfig().getInteger(LOCK_ACQUIRE_NUM_RETRIES_PROP)))
+        .sessionTimeoutMs(lockConfiguration.getConfig().getInteger(ZK_SESSION_TIMEOUT_MS_PROP, DEFAULT_ZK_SESSION_TIMEOUT_MS))
+        .connectionTimeoutMs(lockConfiguration.getConfig().getInteger(ZK_CONNECTION_TIMEOUT_MS_PROP, DEFAULT_ZK_CONNECTION_TIMEOUT_MS))
+        .build();
+    this.curatorFrameworkClient.start();
+  }
+
+  // Only used for testing
+  public ZookeeperBasedLockProvider(
+      final LockConfiguration lockConfiguration, final CuratorFramework curatorFrameworkClient) {
+    checkRequiredProps(lockConfiguration);
+    this.lockConfiguration = lockConfiguration;
+    this.curatorFrameworkClient = curatorFrameworkClient;
+    synchronized (this.curatorFrameworkClient) {
+      if (this.curatorFrameworkClient.getState() != CuratorFrameworkState.STARTED) {
+        this.curatorFrameworkClient.start();
+      }
+    }
+  }
+
+  @Override
+  public boolean tryLock(long time, TimeUnit unit) {
+    LOG.info(generateLogStatement(LockState.ACQUIRING, generateLogSuffixString()));
+    try {
+      acquireLock(time, unit);
+      LOG.info(generateLogStatement(LockState.ACQUIRED, generateLogSuffixString()));
+    } catch (Exception e) {
+      throw new HoodieLockException(generateLogStatement(LockState.FAILED_TO_ACQUIRE, generateLogSuffixString()), e);
+    }
+    return lock != null && lock.isAcquiredInThisProcess();
+  }
+
+  @Override
+  public void unlock() {
+    try {
+      LOG.info(generateLogStatement(LockState.RELEASING, generateLogSuffixString()));
+      if (lock == null || !lock.isAcquiredInThisProcess()) {
+        return;
+      }
+      lock.release();
+      lock = null;
+      LOG.info(generateLogStatement(LockState.RELEASED, generateLogSuffixString()));
+    } catch (Exception e) {
+      throw new HoodieLockException(generateLogStatement(LockState.FAILED_TO_RELEASE, generateLogSuffixString()), e);
+    }
+  }
+
+  @Override
+  public void close() {
+    try {
+      if (lock != null) {
+        lock.release();
+        lock = null;
+      }
+      this.curatorFrameworkClient.close();
+    } catch (Exception e) {
+      LOG.error(generateLogStatement(LockState.FAILED_TO_RELEASE, generateLogSuffixString()));
+    }
+  }
+
+  @Override
+  public InterProcessMutex getLock() {
+    return this.lock;
+  }
+
+  private void acquireLock(long time, TimeUnit unit) throws Exception {
+    ValidationUtils.checkArgument(this.lock == null, LockState.ALREADY_ACQUIRED.name());
```

Comment:
a better validation message, than just the enum name?  (applicable for any such usage)